### PR TITLE
[SofaCombinatorialMaps] dynamic topology into Sofa

### PR DESCRIPTION
**Pull request by untereiner**
Hi guys,

This PR is a proof of concept to show an alternative implementation dynamic topologies and meshes using the CGOGN library.  If you never heard about, CGOGN is a topological modeling kernel (see https://cgogn.github.io) implementing efficiently combinatorial maps to encode the topology and geometrical informations. 

The current implementation is based on implementing new topology into new components.

The PR introduce the following components:  
- VolumeTopologyContainer 
- CMTetrahedralCorotationalFEMForceField 
- MapTetrahedronSetTopologyContainer (wrapper that allows to use "classical" sofa components)

The PR is not supposed to be merged in this state and we welcome any feedback on how to improve and generalize this work.

EDIT: I (damien) changed the description of the PR trying to summary the discussion. 

EDIT 2: 
Steps to compile:
- checkout the branch with a modularized version from  [ (https://github.com/mimesis-inria/sofa/tree/cmtopology_module)](https://github.com/mimesis-inria/sofa/tree/cmtopology_module)
- checkout CGoGN in a folder sofa/modules/SofaCombinatorialMaps/cgogn2
Compile and run the examples in sofa/modules/SofaCombinatorialMaps/examples

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
Wow, thank Lionel for the PR. 
This is a fat one :)

**Comment by untereiner** [comment]
😄 
I do not know how to proceed for you to test it. Maybe I can add test scenes to start somewhere.

**Comment by damienmarchal** [comment]
Hi @untereiner 

From my experience you will have more feedback if you provide more information on the PR description to give other some envy to try it. 

Given the complexity of the PR here is what people passing by may need:
- what the PR does precisely ? 
- what is the existing state on this topic in Sofa (the title speak about dynamic topology...isn't there dynamic topology in Sofa since year) ? 
- how this PR handle the problem, 
- use CGOGN library... what is this CGOGN library ? 
- argument on why the PR is it useful/mandatory/improvement,
- trying to convince people why they should spend time testing it,
- provide some idea on how invasive the PR will be (non breaking at all, breaking the SofaCore API, breaking existing scenes)

**Comment by untereiner** [comment]
Hi guys,

Actually I think this PR should not be merged in this state. It is a proof of concept to show the possibilities.
I just updated a scene that you can use to test the new components.

The scene called "TetrahedralCorotationalFEMForceField" has two nodes. The first one uses 
the components "VolumeTopologyContainer" and "CMTetrahedralCorotationalFEMForceField" that are completely written using cgogn and "MapTetrahedronSetTopologyContainer" a compatibility wrapper that allows to use "classical" sofa components with cgogn.

@damienmarchal concerning your questions:

- what the PR does precisely ?
  - This PR adds a new data structure for handling meshes. This data structure is designed to handle dynamic (with topological changes) meshes.
- what is the existing state on this topic in Sofa (the title speak about dynamic topology...isn't there dynamic topology in Sofa since year) ?
  - I think there is a pipeline for topological changes in Sofa. I do not know much about the efficiency of this pipeline. What I can say (I should provide figures) is that this data structure is as effective as the sofa ones in "static" scenarios and more effective and reliable in "dynamic" scenarios.
- how this PR handle the problem,
  - by changing to a more appropriate data structure
- use CGOGN library... what is this CGOGN library ?
  - a topological modeling kernel (see https://cgogn.github.io). I am not sure you need more information about it.
- argument on why the PR is it useful/mandatory/improvement, trying to convince people why they should spend time testing it,
  - this work has been in progress since 3 years now. His interest has already been debated. I do not want to debate an infinite time on it. If someone is interested in topological changes, it should save him time.
- provide some idea on how invasive the PR will be (non breaking at all, breaking the SofaCore API, breaking existing scenes)
  - depends on the meaning of beaking. It does not change the sofa API for topology classes but it adds a hierarchy of classes in the core with a new API and a compatibility layer with the old one.

**Comment by damienmarchal** [comment]
Thanks for the answers, I moved them to the PR description. 


"This work has been in progress since 3 years now. His interest has already been debated. I do not want to debate an infinite time on it. " 
I searched in the issues and PRs if there was ever any debate on that topic so that I could provide a link in the description but I didn't found one. Maybe you are referring to face to face discussions ? If this was the case maybe writing a summary of these debate (pro/cons) would be very informative and helpful for all of us.  

"depends on the meaning of breaking"
In general we are considering the following:
- a PR is non breaking if its merge will be un-noticeable for people that don't care of the added feature. 
- a PR is breaking the scene if, after the PR, *any* scenes that was working before is not working anymore. 
- a PR is breaking the API if *any* plugin compiling with master is not compiling anymore after the PR. 


Damien

File: SofaKernel/framework/sofa/core/CMakeLists.txt [hunk 1]
```diff
@@ -130,6 +130,13 @@ set(HEADER_FILES
     topology/TopologyChange.h
     topology/TopologyElementHandler.h
     topology/TopologyHandler.h
+    topology/MapTopology.h
+    topology/CMTopologyChange.h
```
Comment (damienmarchal) [inline SofaKernel/framework/sofa/core/CMakeLists.txt]: @untereiner do you think it is possible to have the topology/CM* files put in a plugin so we can do stuff like that:
```xml
<RequiredPlugin name="CGognTopology">
<RequiredPlugin name="CGognForceField">
```

Where the plugin CGognForceField depend on CGOgnTopology ?

File: SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h [hunk 2]
```diff
@@ -252,13 +252,15 @@ class TopologyBarycentricMapper : public BarycentricMapper<In,Out>
     virtual void resize( core::State<Out>* toModel ) = 0;
 
 protected:
-    TopologyBarycentricMapper(core::topology::BaseMeshTopology* fromTopology, topology::PointSetTopologyContainer* toTopology = NULL)
-        : fromTopology(fromTopology), toTopology(toTopology)
+	TopologyBarycentricMapper(core::topology::BaseMeshTopology* fromTopology, topology::PointSetTopologyContainer* toTopology = NULL, bool useRestPosition = false)
+		: fromTopology(fromTopology), toTopology(toTopology), useRestPosition(useRestPosition)
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h]: The change in this file are independent to CM topo and non breaking. So maybe merged through cherry picking in a PR ?

**Comment by damienmarchal** [comment]
After a quick look here are my notes:
- CM is for CombinatorialMap (in case you wonder :)), 

- there are changes on some core mapping that seems to add only minor details as the 'useRestPosition' 
in SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.h
Could this be merged independenty using a fast path. 

- could the the extlibs/NGOGN should be removed to help in readbility of the PR

- is it possible that the new topology layer (from sofa/core/topology and SofaBaseTopology) to be moved out of the core in a module/plugin CMTopology and be part of the same cmtopology namespace. 

- could the new CMforcefields moved to a module/plugins CMMechanics ? 

- why there is files that appears like new files but does not seems to use the cmtopology like ?:
modules/SofaMiscFem/ElementFEMForceField.cpp
modules/SofaMiscMapping/BeamMapping.h
is this a mistake or do I mis-understood something ? 

DM. 
NOTE: I will edit my notes if after more reading I found some note are not relevant.

**Comment by guparan** [comment]
I'll try to take some time to refactor this.

**Comment by guparan** [comment]
Here is a rebased + cleaned + updated version without CGoGN: https://github.com/sofa-framework/sofa/compare/master...guparan:cmtopology_rebased_cleaned

To be mixed with this extracted CGoGN: https://github.com/guparan/sofa_cgogn

Please tell me if it works :-)

**Comment by damienmarchal** [comment]
75 files changes...instead of 1300 :) so much better. 
Thanks Guillaume.

**Comment by untereiner** [comment]
Thank you @guparan! 

What I see first is:
- [ ] ~~the files modules/SofaMiscFem/HexaFEMForceField.* should be renamed with CM prefix~~
- [x] the files modules/SofaMiscMapping/BeamMapping.* should be erased (merged from another branch and unrelated to combinatorial topologies ) 
- [x] the changes to modules/SofaGeneralSimpleFem/BeamFEMForceField.* should be erased (merged from another branch, juste a wip and unrelated to combinatorial topologies)
- [x] And as said by @damienmarchal, the changes related to SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.* should be extracted to another branch (unrelated to combinatorial topologies) 
- [x] the modifications in the files modules/SofaGeneralSimpleFem/TetrahedralCorotationalFEMForceField.* should be erased (test work of one of our former trainee)
- [x] The file modules/SofaMiscMapping/BarycentricMappingRigid.h has modifications unrelated to combinatorial topologies should be extracted to another branch too

These modifications should decrease the number of changed files a bit more

**Comment by guparan** [comment]
>  the files modules/SofaMiscFem/HexaFEMForceField.* should be renamed with CM prefix

Removed HexaFEMForceField and ElementFEMForceField instead.

**Comment by untereiner** [comment]
With the help of @guparan with made a few :) changes to this branch. To summarize:
- we made all changes I listed above to remove dead or unrelated code
- since the messages were not very helpfull, all the remaining commits were squashed in a single one which is the contribution.

cgogn is for now an external dependency and has to be checkouted manually in the extlib directory in order to compile the branch. We have to discuss a way to do it automatically (maybe an ExternalProject_Add ?). After a quick check, this proposal is compatible with the last commit of the cgogn devel branch.

**Comment by untereiner** [comment]
[ci-build][with-scene-tests]

**Comment by untereiner** [comment]
Hi everyone.

I just created a new branch (https://github.com/mimesis-inria/sofa/tree/cmtopology_module) where I put everything into a module called "SofaCombinatorialMaps" instead of using different existing sofa modules.
To test it, you just have to put the cgogn repository that @guparan extrated earlier into a sofa/modules/SofaCombinatorialMaps/cgogn2 directory and it should compile :crossed_fingers:.
I am able to launch examples/CMTetrahedralCorotationalFEMForceField.scn

**Comment by guparan** [comment]
This PR now contains the "module" version of cmtopology.
Reviews are welcome :+1:

**Comment by guparan** [comment]
I did some cleaning on the PR but still need some help to make it compile.
If someone could check it out that would be great :1st_place_medal:

**Comment by damienmarchal** [comment]
I was looking at he "stalled" PR,

And was wondering about this one, it contains a lot of valuable work and it seems to be connected to the cleanig & refactoring work on topologies started by @epernod. 

@guparan , @epernod , @untereiner,  @StephaneCotin
what are your suggestion on how to move forward and have that work "not lost" ?

**Comment by untereiner** [comment]
There is only a windows compilation problem if I remember correctly. If someone can solve this issue I think It will be mergeable asap.

**Pull request closed without merging**